File: sexy_primes.pl
sexy_prime_group(1, N, _, [N]):-
    is_prime(N),
    !.
sexy_prime_group(Size, N, Limit, [N|Group]):-
    is_prime(N),
    N1 is N + 6,
    N1 =< Limit,
    S1 is Size - 1,
    sexy_prime_group(S1, N1, Limit, Group).
  
sexy_prime_groups(_, [], _, []):-
    !.
sexy_prime_groups(Size, [P|Primes], Limit, [Group|Groups]):-
    sexy_prime_group(Size, P, Limit, Group),
    !,
    sexy_prime_groups(Size, Primes, Limit, Groups).
sexy_prime_groups(Size, [_|Primes], Limit, Groups):-
    sexy_prime_groups(Size, Primes, Limit, Groups).

print_sexy_prime_groups(Size, Primes, Limit):-
    sexy_prime_groups(Size, Primes, Limit, Groups),
    length(Groups, Len),
    writef('Number of groups of size %t is %t\n', [Size, Len]),
    last_n(Groups, 5, Len, Last, Last_len),
    writef('Last %t groups of size %t: %t\n\n', [Last_len, Size, Last]).

last_n([], _, L, [], L):-!.
last_n([_|List], Max, Length, Last, Last_len):-
    Max < Length,
    !,
    Len1 is Length - 1,
    last_n(List, Max, Len1, Last, Last_len).
last_n([E|List], Max, Length, [E|Last], Last_len):-
    last_n(List, Max, Length, Last, Last_len).

unsexy_prime(P):-
    is_prime(P),
    P1 is P + 6,
    \+is_prime(P1),
    P2 is P - 6,
    \+is_prime(P2).

unsexy_primes(Limit, [2|P]):-
    unsexy_primes(3, Limit, P).

unsexy_primes(From, To, []):-
    From > To,
    !.
unsexy_primes(From, To, [From|Rest]):-
    unsexy_prime(From),
    !,
    Next is From + 2,
    unsexy_primes(Next, To, Rest).
unsexy_primes(From, To, Rest):-
    Next is From + 2,
    unsexy_primes(Next, To, Rest).

main(Limit):-
    Max is Limit + 6,
    find_prime_numbers(Max),
    findall(P, (is_prime(P), P =< Limit), Primes),
    print_sexy_prime_groups(2, Primes, Limit),
    print_sexy_prime_groups(3, Primes, Limit),
    print_sexy_prime_groups(4, Primes, Limit),
    print_sexy_prime_groups(5, Primes, Limit),
    unsexy_primes(Limit, Unsexy),
    length(Unsexy, Count),
    writef('Number of unsexy primes is %t\n', [Count]),
    last_n(Unsexy, 10, Count, Last10, _),
    writef('Last 10 unsexy primes: %t', [Last10]).

main:-
    main(1000035).

File: sexy_primes/sexy_primes.cpp
#include <array>
#include <iostream>
#include <vector>
#include "../library/circular_buffer.h"
#include "../library/sieve_of_eratosthenes.h"

int main() {
    using std::cout;
    using std::vector;

    const int max = 1000035;
    const int max_group_size = 5;
    const int diff = 6;
    const int array_size = max + diff;
    const int max_groups = 5;
    const int max_unsexy = 10;

    // Use Sieve of Eratosthenes to find prime numbers up to max
    sieve_of_eratosthenes sieve(array_size);

    std::array<int, max_group_size> group_count{0};
    vector<circular_buffer<vector<int>>> groups;
    for (size_t i = 0; i < max_group_size; ++i)
        groups.emplace_back(max_groups);
    int unsexy_count = 0;
    circular_buffer<int> unsexy_primes(max_unsexy);

    for (int p = 2; p < max; ) {
        if (!sieve.is_prime(p + diff) && (p - diff < 2 || !sieve.is_prime(p - diff))) {
            // if p + diff and p - diff aren't prime then p can't be sexy
            ++unsexy_count;
            unsexy_primes.push_back(p);
        } else {
            // find the groups of sexy primes that begin with p
            int group_size = 1;
            vector<int> group;
            group.push_back(p);
            for (int i = 1; i < max_group_size; ++i) {
                int next_p = p + i * diff;
                if (next_p >= max || !sieve.is_prime(next_p))
                    break;
                ++group_size;
                group.push_back(next_p);
                groups[group_size - 1].push_back(group);
            }
            for (int i = 1; i < group_size; ++i)
                ++group_count[i];
        }
        // skip to next prime number
        ++p;
        while (p < max && !sieve.is_prime(p))
            ++p;
    }

    for (int size = 1; size < max_group_size; ++size) {
        cout << "number of groups of size " << size + 1 << " is " << group_count[size] << '\n';
        cout << "last " << groups[size].size() << " groups of size " << size + 1 << ":";
        for (const vector<int>& group : groups[size]) {
            cout << " (";
            for (size_t i = 0; i < group.size(); ++i) {
                if (i > 0)
                    cout << ' ';
                cout << group[i];
            }
            cout << ")";
        }
        cout << "\n\n";
    }
    cout << "number of unsexy primes is " << unsexy_count << '\n';
    cout << "last " << unsexy_primes.size() << " unsexy primes:";
    for (int prime : unsexy_primes)
        cout << ' ' << prime;
    cout << '\n';
    return 0;
}
